src/v/cluster/shard_placement_table.cc

```diff
@@ -171,82 +172,69 @@ ss::future<> shard_placement_table::initialize(
               }
           });
     }
+
+    if (!_ntp2entry.empty()) {
+        _cur_shard_revision += 1;
+    }
 }
 
 ss::future<> shard_placement_table::set_target(
   const model::ntp& ntp,
   std::optional<shard_placement_target> target,
-  model::shard_revision_id shard_rev,
   shard_callback_t shard_callback) {
     vassert(
       ss::this_shard_id() == assignment_shard_id,
       "method can only be invoked on shard {}",
       assignment_shard_id);
 
-    auto units = co_await _mtx.get_units();
-
-    // 1. update node-wide map
-
-    std::optional<ss::shard_id> prev;
-    bool is_initial = false;
-    if (target) {
-        auto [it, inserted] = _ntp2target.try_emplace(ntp, *target);
-        if (inserted) {
-            vlog(
-              clusterlog.trace,
-              "[{}] insert target: {}, shard_rev: {}",
-              ntp,
-              target,
-              shard_rev);
-            is_initial = true;
-        } else {
-            if (it->second == *target) {
-                vlog(
-                  clusterlog.trace,
-                  "[{}] modify target no-op, cur: {}, shard_rev: {}",
-                  ntp,
-                  it->second,
-                  shard_rev);
-                co_return;
-            }
-
-            vlog(
-              clusterlog.trace,
-              "[{}] modify target: {} -> {}, shard_rev: {}",
-              ntp,
-              it->second,
-              target,
-              shard_rev);
-            is_initial = it->second.log_revision != target->log_revision;
-            if (it->second.shard != target->shard) {
-                prev = it->second.shard;
-            }
-            it->second = *target;
-        }
-    } else {
-        auto prev_it = _ntp2target.find(ntp);
-        if (prev_it == _ntp2target.end()) {
-            vlog(
-              clusterlog.trace,
-              "[{}] remove target no-op, shard_rev: {}",
-              ntp,
-              shard_rev);
-            co_return;
-        }
+    if (!target && !_ntp2entry.contains(ntp)) {
+        co_return;
+    }
 
+    auto entry_it = _ntp2entry.try_emplace(ntp).first;
+    if (!entry_it->second) {
+        entry_it->second = std::make_unique<entry_t>();
+    }
+    entry_t& entry = *entry_it->second;
+
+    auto release_units = ss::defer(
+      [&, units = co_await entry.mtx.get_units()]() mutable {
+          bool had_waiters = entry.mtx.waiters() > 0;
```

Comment:
@ztlpn even tho entyr_it->second is a heap unique_ptr (so not worried about container resharding), is there a worry that entry_it could get removed during a scheduling event and this entry reference to the heap entry would be freed?

Reply:
This shouldn't happen as the only place where we remove `entry_it` is this method, and another invocation of it would block on this mutex before doing anything.